vpr/src/pack/cluster_legalizer.h

```diff
@@ -0,0 +1,474 @@
+ *          break;
+ *
+ * // Do the expensive check once all molecules are in.
+ * if (!legalizer.check_cluster_legality(new_cluster_id))
+ *      // Destroy the illegal cluster.
+ *      legalizer.destroy_cluster(new_cluster_id);
+ *      legalizer.compress();
+ *      // Handle how to try again (maybe use FULL strategy).
+ *
+ * 2) FULL Legalization Strategy Example:
+ * This strategy will fully route the internal connections of the clusters for
+ * each molecule added. This is much more expensive to run; however, will ensure
+ * that the cluster is fully legalized while it is being created. An example
+ * of how to sue this strategy would look something like this:
+ *
+ * Clusterlegalizer legalizer(...);
+ *
+ * std::tie(pack_result, new_cluster_id) = legalizer.start_new_cluster(seed_mol,
+ *                                                                     cluster_type,
+ *                                                                     mode);
+ * for mol in molecules_to_add:
+ *      // Do the expensive check for each molecule added.
+ *      status = legalizer.add_mol_to_cluster(mol, new_cluster_id);
+ *      if (status != e_block_pack_status::BLK_PASSED)
+ *          break;
+ *
+ * // new_cluster_id now contains a fully legalized cluster.
+ */
+class ClusterLegalizer {
+public:
+    // Iterator for the legalization cluster IDs
+    typedef typename vtr::vector_map<LegalizationClusterId, LegalizationClusterId>::const_iterator cluster_iterator;
+
+    // Range for the legalization cluster IDs
+    typedef typename vtr::Range<cluster_iterator> cluster_range;
+
+private:
+
+    /*
+     * @brief Helper method that tries to pack the given molecule into a cluster.
+     *
+     * This method runs all the legality checks specified by the legalization
+     * strategy. If the molecule can be packed into the cluster, it will insert
+     * it into the cluster.
+     *
+     *  @param molecule                 The molecule to insert into the cluster.
+     *  @param cluster                  The cluster to try to insert the molecule into.
+     *  @param cluster_id               The ID of the cluster.
+     *  @param max_external_pin_util    The max external pin utilization for a
+     *                                  cluster of this type.
+     */
+    e_block_pack_status try_pack_molecule(t_pack_molecule* molecule,
+                                          LegalizationCluster& cluster,
+                                          LegalizationClusterId cluster_id,
+                                          const t_ext_pin_util& max_external_pin_util);
+
+public:
+
+    // Explicitly deleted default constructor. Need to use other constructor to
+    // initialize state correctly.
+    ClusterLegalizer() = delete;
+
+    /*
+     * @brief Initialize the ClusterLegalizer class.
+     *
+     * Allocates internal state.
+     */
+    ClusterLegalizer(const AtomNetlist& atom_netlist,
+                     const Prepacker& t_prepacker,
```

Comment:
You have a good point. Ill see what I can do.

Reply:
I opted to use the same name, and use the "this->" syntax when it was confusing. I personally do not like the "_<member_variable>" syntax since I think "_<variable_name>" may be reserved by some compilers.

Reply:
Nevermind, the compiler was giving warnings. I will move to the `<member_variable>_` syntax.